*: enables elastic CPU limiter for all users of ExportRequest

adityamaru opened the pull request:
Previously, there was a strange coupling between the elastic CPU
limiter and the `header.TargetBytes` DistSender limit set on each
ExportRequest. Even if a request was preempted on exhausting its
allotted CPU tokens, it would only return from kvserver by virtue
of its `header.TargetBytes` being set to a non-zero value. Out of the
four users of ExportRequest, only backup set this field to a sentinel
value of 1 to limit the number of SSTs we send back in an ExportResponse.
The remaining callers of ExportRequest would not return from the kvserver.
Instead they would evaluate the request from the resume key immediately,
not giving the scheduler a chance to take the goroutine off CPU.

This change breaks this coupling by introducing a `resumeInfo` object that
indicates whether the resumption was because we were over our CPU limit. If
it was, we return an ExportResponse with our progress so far. This change
shifts the burden of handling pagination to the client. This seems better than
having the server sleep or wait around until its CPU tokens are replenished
as the client would be left wondering why a request is taking so long.

To that effect this change adds pagination support to the other callers of
ExportRequest. Note, we do not set `SplitMidKey` at these other callsites
yet. Thus, all pagination will happen at key boundaries in the ExportRequest.
A follow-up will add support for `SplitMidKey` to these callers.

Informs: https://github.com/cockroachdb/cockroach/issues/96684

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96691)
<!-- Reviewable:end -->

adityamaru committed:
*: enables elastic CPU limiter for all users of ExportRequest

Previously, there was a strange coupling between the elastic CPU
limiter and the `header.TargetBytes` DistSender limit set on each
ExportRequest. Even if a request was preempted on exhausting its
allotted CPU tokens, it would only return from kvserver by virtue
of its `header.TargetBytes` being set to a non-zero value. Out of the
four users of ExportRequest, only backup set this field to a sentinel
value of 1 to limit the number of SSTs we send back in an ExportResponse.
The remaining callers of ExportRequest would not return from the kvserver.
Instead they would evaluate the request from the resume key immediately,
not giving the scheduler a chance to take the goroutine off CPU.

This change breaks this coupling by introducing a `resumeInfo` object that
indicates whether the resumption was because we were over our CPU limit. If
it was, we return an ExportResponse with our progress so far. This change
shifts the burden of handling pagination to the client. This seems better than
having the server sleep or wait around until its CPU tokens are replenished
as the client would be left wondering why a request is taking so long.

To that effect this change adds pagination support to the other callers of
ExportRequest. Note, we do not set `SplitMidKey` at these other callsites
yet. Thus, all pagination will happen at key boundaries in the ExportRequest.
A follow-up will add support for `SplitMidKey` to these callers.

Informs: #96684

Release note: None

adityamaru commented:
I was playing around with the test coverage tool and we have a pretty good coverage for `mvcc.go` and `cmd_export.go`.

<img width="1661" alt="Screenshot 2023-02-09 at 11 16 46 AM" src="https://user-images.githubusercontent.com/13837382/217872778-fb5a8d57-3598-496a-9417-17b53fbeafed.png">

<img width="1660" alt="Screenshot 2023-02-09 at 11 32 04 AM" src="https://user-images.githubusercontent.com/13837382/217877342-83b05b89-ffa8-49d1-8e3a-3490383b9fb7.png">

adityamaru commented:
friendly ping @stevendanna / @irfansharif , I have a couple of export related diffs queued up behind this so just wanted to put this on your radar. Thanks!

stevendanna commented on pkg/sql/catalog/lease/lease.go:
```diff
@@ -254,85 +254,103 @@ func getDescriptorsFromStoreForInterval(
-		}(); err != nil {
-			return nil, err
 		}
+		log.Infof(ctx, "export response %s", descriptorKey.String())
```
How noisy will this be in a production cluster?

stevendanna commented on pkg/kv/kvserver/batcheval/cmd_export.go:
```diff
@@ -249,21 +277,39 @@ func evalExport(
+					Key:    resumeInfo.ResumeKey.Key,
+					EndKey: args.EndKey,
+				}
+				// TODO(during review): Do we want to add another resume reason
+				// specifically for CPU preemption.
+				reply.ResumeReason = roachpb.RESUME_UNKNOWN
```
I didn't spot anywhere that a new reasons would cause a problem. Very few locations seem to even read the reason.  My vote would be to add a reason since we do know the reason here.

adityamaru replied:
done!

stevendanna commented on pkg/kv/kvserver/batcheval/cmd_export.go:
```diff
@@ -225,18 +225,46 @@ func evalExport(
+					reply.ResumeReason = roachpb.RESUME_UNKNOWN
+					break
+				} else {
+					// We should never come here. There should be no condition aside from
+					// resource constraints that results in an early exit without
+					// exporting any data. Regardless, if we have a resumeKey we
+					// immediately retry the ExportRequest from that key and timestamp
+					// onwards.
```
Depending on our level of confident here, perhaps we could return an error in this case only in non-release builds and see if we ever trigger it in CI.

stevendanna reviewed (approved):
Overall this looks good to me.  I've left some possible suggestions but none of them are blocking in my opinion.

adityamaru commented:
flake is https://github.com/cockroachdb/cockroach/issues/97320

bors r=stevendanna
